diff --git a/include/class_loader/class_loader_core.hpp b/include/class_loader/class_loader_core.hpp
index bfb17fe..40c1138 100644
--- a/include/class_loader/class_loader_core.hpp
+++ b/include/class_loader/class_loader_core.hpp
@@ -285,7 +285,7 @@ Base * createInstance(const std::string & derived_class_name, ClassLoader * load
   getPluginBaseToFactoryMapMapMutex().lock();
   FactoryMap & factoryMap = getFactoryMapForBaseClass<Base>();
   if (factoryMap.find(derived_class_name) != factoryMap.end()) {
-    factory = dynamic_cast<impl::AbstractMetaObject<Base> *>(factoryMap[derived_class_name]);
+    factory = static_cast<impl::AbstractMetaObject<Base> *>(factoryMap[derived_class_name]);
   } else {
     CONSOLE_BRIDGE_logError(
       "class_loader.impl: No metaobject exists for class type %s.", derived_class_name.c_str());
