diff --git a/orocos_kdl_vendor/CMakeLists.txt b/orocos_kdl_vendor/CMakeLists.txt
index 655e562..b953c94 100644
--- a/orocos_kdl_vendor/CMakeLists.txt
+++ b/orocos_kdl_vendor/CMakeLists.txt
@@ -18,6 +18,9 @@ macro(build_orocos_kdl)
   if(DEFINED CMAKE_BUILD_TYPE)
     list(APPEND extra_cmake_args -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE})
   endif()
+  if(DEFINED EIGEN3_INCLUDE_DIR)
+    list(APPEND extra_cmake_args "-DEIGEN3_INCLUDE_DIR=${EIGEN3_INCLUDE_DIR}")
+  endif()
   if(DEFINED CMAKE_TOOLCHAIN_FILE)
     list(APPEND extra_cmake_args "-DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}")
     if(ANDROID)
@@ -64,6 +67,8 @@ macro(build_orocos_kdl)
     SOURCE_SUBDIR orocos_kdl
     CMAKE_ARGS
       -DCMAKE_INSTALL_PREFIX=${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}_install
+      -DCXX_STANDARD=17
+      -DCMAKE_CXX_STANDARD=17
       ${extra_cmake_args}
     PATCH_COMMAND
       ${CMAKE_COMMAND} -E chdir <SOURCE_DIR> git apply -p1 --ignore-space-change --whitespace=nowarn
diff --git a/python_orocos_kdl_vendor/CMakeLists.txt b/python_orocos_kdl_vendor/CMakeLists.txt
index 48554fd..36bae2f 100644
--- a/python_orocos_kdl_vendor/CMakeLists.txt
+++ b/python_orocos_kdl_vendor/CMakeLists.txt
@@ -4,6 +4,9 @@ project(python_orocos_kdl_vendor)
 
 find_package(ament_cmake REQUIRED)
 
+set(CMAKE_CXX_STANDARD 17)
+set(CXX_STANDARD 17)
+
 option(FORCE_BUILD_VENDOR_PKG
   "Build python_orocos_kdl from source, even if system-installed package is available"
   OFF)
