diff --git a/rmw_cyclonedds_cpp/src/TypeSupport.hpp b/rmw_cyclonedds_cpp/src/TypeSupport.hpp
index be07c9e..413f21a 100644
--- a/rmw_cyclonedds_cpp/src/TypeSupport.hpp
+++ b/rmw_cyclonedds_cpp/src/TypeSupport.hpp
@@ -78,7 +78,7 @@ struct StringHelper<rosidl_typesupport_introspection_c__MessageMembers>
 
   static void assign(cycdeser & deser, void * field)
   {
-    std::string str;
+    std::pmr::string str;
     deser >> str;
     rosidl_runtime_c__String * c_str = static_cast<rosidl_runtime_c__String *>(field);
     rosidl_runtime_c__String__assign(c_str, str.c_str());
@@ -89,16 +89,16 @@ struct StringHelper<rosidl_typesupport_introspection_c__MessageMembers>
 template<>
 struct StringHelper<rosidl_typesupport_introspection_cpp::MessageMembers>
 {
-  using type = std::string;
+  using type = std::pmr::string;
 
-  static std::string & convert_to_std_string(void * data)
+  static type & convert_to_std_string(void * data)
   {
-    return *(static_cast<std::string *>(data));
+    return *(static_cast<type *>(data));
   }
 
   static void assign(cycdeser & deser, void * field)
   {
-    std::string & str = *(std::string *)field;
+    type & str = *(type *)field;
     deser >> str;
   }
 };
diff --git a/rmw_cyclonedds_cpp/src/TypeSupport2.hpp b/rmw_cyclonedds_cpp/src/TypeSupport2.hpp
index b4b890d..99f8d3b 100644
--- a/rmw_cyclonedds_cpp/src/TypeSupport2.hpp
+++ b/rmw_cyclonedds_cpp/src/TypeSupport2.hpp
@@ -437,7 +437,7 @@ public:
 class ROSIDLCPP_StringValueType : public U8StringValueType
 {
 public:
-  using type = std::string;
+  using type = std::pmr::string;
 
   TypedSpan<const char_traits::char_type> data(const void * ptr) const override
   {
diff --git a/rmw_cyclonedds_cpp/src/TypeSupport_impl.hpp b/rmw_cyclonedds_cpp/src/TypeSupport_impl.hpp
index 736e088..3c28806 100644
--- a/rmw_cyclonedds_cpp/src/TypeSupport_impl.hpp
+++ b/rmw_cyclonedds_cpp/src/TypeSupport_impl.hpp
@@ -113,39 +113,39 @@ void deserialize_field(
   } else if (member->array_size_ && !member->is_upper_bound_) {
     deser.deserializeA(static_cast<T *>(field), member->array_size_);
   } else {
-    auto & vector = *reinterpret_cast<std::vector<T> *>(field);
+    auto & vector = *reinterpret_cast<std::pmr::vector<T> *>(field);
     deser >> vector;
   }
 }
 
 template<>
-inline void deserialize_field<std::string>(
+inline void deserialize_field<std::pmr::string>(
   const rosidl_typesupport_introspection_cpp::MessageMember * member,
   void * field,
   cycdeser & deser)
 {
   if (!member->is_array_) {
-    deser >> *static_cast<std::string *>(field);
+    deser >> *static_cast<std::pmr::string *>(field);
   } else if (member->array_size_ && !member->is_upper_bound_) {
-    std::string * array = static_cast<std::string *>(field);
+    std::pmr::string * array = static_cast<std::pmr::string *>(field);
     deser.deserializeA(array, member->array_size_);
   } else {
-    auto & vector = *reinterpret_cast<std::vector<std::string> *>(field);
+    auto & vector = *reinterpret_cast<std::pmr::vector<std::pmr::string> *>(field);
     deser >> vector;
   }
 }
 
 template<>
-inline void deserialize_field<std::wstring>(
+inline void deserialize_field<std::pmr::wstring>(
   const rosidl_typesupport_introspection_cpp::MessageMember * member,
   void * field,
   cycdeser & deser)
 {
-  std::wstring wstr;
+  std::pmr::wstring wstr;
   if (!member->is_array_) {
     deser >> wstr;
     wstring_to_u16string(
-      wstr, *static_cast<std::u16string *>(field));
+      wstr, *static_cast<std::pmr::u16string *>(field));
   } else {
     uint32_t size;
     if (member->array_size_ && !member->is_upper_bound_) {
@@ -156,7 +156,7 @@ inline void deserialize_field<std::wstring>(
     }
     for (size_t i = 0; i < size; ++i) {
       void * element = member->get_function(field, i);
-      auto u16str = static_cast<std::u16string *>(element);
+      auto u16str = static_cast<std::pmr::u16string *>(element);
       deser >> wstr;
       wstring_to_u16string(wstr, *u16str);
     }
@@ -185,7 +185,7 @@ void deserialize_field(
 }
 
 template<>
-inline void deserialize_field<std::string>(
+inline void deserialize_field<std::pmr::string>(
   const rosidl_typesupport_introspection_c__MessageMember * member,
   void * field,
   cycdeser & deser)
@@ -198,7 +198,7 @@ inline void deserialize_field<std::string>(
       auto deser_field = static_cast<rosidl_runtime_c__String *>(field);
       // tmpstring is defined here and not below to avoid
       // memory allocation in every iteration of the for loop
-      std::string tmpstring;
+      std::pmr::string tmpstring;
       for (size_t i = 0; i < member->array_size_; ++i) {
         deser.deserialize(tmpstring);
         if (!rosidl_runtime_c__String__assign(&deser_field[i], tmpstring.c_str())) {
@@ -206,7 +206,7 @@ inline void deserialize_field<std::string>(
         }
       }
     } else {
-      std::vector<std::string> cpp_string_vector;
+      std::pmr::vector<std::pmr::string> cpp_string_vector;
       deser >> cpp_string_vector;
 
       auto & string_array_field = *reinterpret_cast<rosidl_runtime_c__String__Sequence *>(field);
@@ -230,12 +230,12 @@ inline void deserialize_field<std::string>(
 }
 
 template<>
-inline void deserialize_field<std::wstring>(
+inline void deserialize_field<std::pmr::wstring>(
   const rosidl_typesupport_introspection_c__MessageMember * member,
   void * field,
   cycdeser & deser)
 {
-  std::wstring wstr;
+  std::pmr::wstring wstr;
   if (!member->is_array_) {
     deser >> wstr;
     wstring_to_u16string(
@@ -307,10 +307,10 @@ bool TypeSupport<MembersType>::deserializeROSmessage(
         deserialize_field<uint64_t>(member, field, deser);
         break;
       case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
-        deserialize_field<std::string>(member, field, deser);
+        deserialize_field<std::pmr::string>(member, field, deser);
         break;
       case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_WSTRING:
-        deserialize_field<std::wstring>(member, field, deser);
+        deserialize_field<std::pmr::wstring>(member, field, deser);
         break;
       case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE:
         {
@@ -412,10 +412,10 @@ bool TypeSupport<MembersType>::printROSmessage(
         {uint64_t dummy; print_field(member, deser, dummy);}
         break;
       case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_STRING:
-        {std::string dummy; print_field(member, deser, dummy);}
+        {std::pmr::string dummy; print_field(member, deser, dummy);}
         break;
       case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_WSTRING:
-        {std::wstring dummy; print_field(member, deser, dummy);}
+        {std::pmr::wstring dummy; print_field(member, deser, dummy);}
         break;
       case ::rosidl_typesupport_introspection_cpp::ROS_TYPE_MESSAGE:
         {
diff --git a/rmw_cyclonedds_cpp/src/serdes.hpp b/rmw_cyclonedds_cpp/src/serdes.hpp
index 9973572..2270ee0 100644
--- a/rmw_cyclonedds_cpp/src/serdes.hpp
+++ b/rmw_cyclonedds_cpp/src/serdes.hpp
@@ -110,10 +110,10 @@ public:
   inline cycdeser & operator>>(uint64_t & x) {deserialize(x); return *this;}
   inline cycdeser & operator>>(float & x) {deserialize(x); return *this;}
   inline cycdeser & operator>>(double & x) {deserialize(x); return *this;}
-  inline cycdeser & operator>>(std::string & x) {deserialize(x); return *this;}
-  inline cycdeser & operator>>(std::wstring & x) {deserialize(x); return *this;}
-  template<class T>
-  inline cycdeser & operator>>(std::vector<T> & x) {deserialize(x); return *this;}
+  inline cycdeser & operator>>(std::pmr::string & x) {deserialize(x); return *this;}
+  inline cycdeser & operator>>(std::pmr::wstring & x) {deserialize(x); return *this;}
+  template<typename T>
+  inline cycdeser & operator>>(std::pmr::vector<T> & x) {deserialize(x); return *this;}
   template<class T, size_t S>
   inline cycdeser & operator>>(std::array<T, S> & x) {deserialize(x); return *this;}
 
@@ -155,22 +155,22 @@ public:
     validate_size(sz, el_sz);
     return sz;
   }
-  inline void deserialize(std::string & x)
+  inline void deserialize(std::pmr::string & x)
   {
     const uint32_t sz = deserialize_len(sizeof(char));
     if (sz == 0) {
-      x = std::string("");
+      x = std::pmr::string("");
     } else {
       validate_str(sz);
-      x = std::string(data + pos, sz - 1);
+      x = std::pmr::string(data + pos, sz - 1);
     }
     pos += sz;
   }
-  inline void deserialize(std::wstring & x)
+  inline void deserialize(std::pmr::wstring & x)
   {
     const uint32_t sz = deserialize_len(sizeof(wchar_t));
     // wstring is not null-terminated in cdr
-    x = std::wstring(reinterpret_cast<const wchar_t *>(data + pos), sz);
+    x = std::pmr::wstring(reinterpret_cast<const wchar_t *>(data + pos), sz);
     pos += sz * sizeof(wchar_t);
   }
 
@@ -218,14 +218,14 @@ public:
     for (size_t i = 0; i < cnt; i++) {deserialize(x[i]);}
   }
 
-  template<class T>
-  inline void deserialize(std::vector<T> & x)
+  template<typename T>
+  inline void deserialize(std::pmr::vector<T> & x)
   {
     const uint32_t sz = deserialize_len(1);
     x.resize(sz);
     deserializeA(x.data(), sz);
   }
-  inline void deserialize(std::vector<bool> & x)
+  inline void deserialize(std::pmr::vector<bool> & x)
   {
     const uint32_t sz = deserialize_len(sizeof(unsigned char));
     x.resize(sz);
@@ -264,10 +264,10 @@ public:
   inline cycprint & operator>>(uint64_t & x) {print(x); return *this;}
   inline cycprint & operator>>(float & x) {print(x); return *this;}
   inline cycprint & operator>>(double & x) {print(x); return *this;}
-  inline cycprint & operator>>(std::string & x) {print(x); return *this;}
-  inline cycprint & operator>>(std::wstring & x) {print(x); return *this;}
-  template<class T>
-  inline cycprint & operator>>(std::vector<T> & x) {print(x); return *this;}
+  inline cycprint & operator>>(std::pmr::string & x) {print(x); return *this;}
+  inline cycprint & operator>>(std::pmr::wstring & x) {print(x); return *this;}
+  template<typename T>
+  inline cycprint & operator>>(std::pmr::vector<T> & x) {print(x); return *this;}
   template<class T, size_t S>
   inline cycprint & operator>>(std::array<T, S> & x) {print(x); return *this;}
 
@@ -329,7 +329,7 @@ public:
     validate_size(sz, el_sz);
     return sz;
   }
-  inline void print(std::string & x)
+  inline void print(std::pmr::string & x)
   {
     const uint32_t sz = get_len(sizeof(char));
     validate_str(sz);
@@ -338,11 +338,11 @@ public:
     prtf(&buf, &bufsize, "\"%*.*s\"", len, len, static_cast<const char *>(data + pos));
     pos += sz;
   }
-  inline void print(std::wstring & x)
+  inline void print(std::pmr::wstring & x)
   {
     const uint32_t sz = get_len(sizeof(wchar_t));
     // wstring is not null-terminated in cdr
-    x = std::wstring(reinterpret_cast<const wchar_t *>(data + pos), sz);
+    x = std::pmr::wstring(reinterpret_cast<const wchar_t *>(data + pos), sz);
     prtf(&buf, &bufsize, "\"%ls\"", x.c_str());
     pos += sz * sizeof(wchar_t);
   }
@@ -359,7 +359,7 @@ public:
   }
 
   template<class T>
-  inline void print(std::vector<T> & x)
+  inline void print(std::pmr::vector<T> & x)
   {
     const uint32_t sz = get_len(1);
     printA(x.data(), sz);
diff --git a/rmw_cyclonedds_cpp/src/u16string.cpp b/rmw_cyclonedds_cpp/src/u16string.cpp
index 562041a..f5dd5fa 100644
--- a/rmw_cyclonedds_cpp/src/u16string.cpp
+++ b/rmw_cyclonedds_cpp/src/u16string.cpp
@@ -18,7 +18,7 @@
 namespace rmw_cyclonedds_cpp
 {
 
-void u16string_to_wstring(const std::u16string & u16str, std::wstring & wstr)
+void u16string_to_wstring(const std::pmr::u16string & u16str, std::pmr::wstring & wstr)
 {
   wstr.resize(u16str.size());
   for (size_t i = 0; i < u16str.size(); ++i) {
@@ -26,7 +26,7 @@ void u16string_to_wstring(const std::u16string & u16str, std::wstring & wstr)
   }
 }
 
-bool wstring_to_u16string(const std::wstring & wstr, std::u16string & u16str)
+bool wstring_to_u16string(const std::pmr::wstring & wstr, std::pmr::u16string & u16str)
 {
   try {
     u16str.resize(wstr.size());
@@ -39,7 +39,7 @@ bool wstring_to_u16string(const std::wstring & wstr, std::u16string & u16str)
   return true;
 }
 
-void u16string_to_wstring(const rosidl_runtime_c__U16String & u16str, std::wstring & wstr)
+void u16string_to_wstring(const rosidl_runtime_c__U16String & u16str, std::pmr::wstring & wstr)
 {
   wstr.resize(u16str.size);
   for (size_t i = 0; i < u16str.size; ++i) {
@@ -47,7 +47,7 @@ void u16string_to_wstring(const rosidl_runtime_c__U16String & u16str, std::wstri
   }
 }
 
-bool wstring_to_u16string(const std::wstring & wstr, rosidl_runtime_c__U16String & u16str)
+bool wstring_to_u16string(const std::pmr::wstring & wstr, rosidl_runtime_c__U16String & u16str)
 {
   bool succeeded = rosidl_runtime_c__U16String__resize(&u16str, wstr.size());
   if (!succeeded) {
diff --git a/rmw_cyclonedds_cpp/src/u16string.hpp b/rmw_cyclonedds_cpp/src/u16string.hpp
index 4019aa8..3bbbbcb 100644
--- a/rmw_cyclonedds_cpp/src/u16string.hpp
+++ b/rmw_cyclonedds_cpp/src/u16string.hpp
@@ -16,20 +16,21 @@
 #define U16STRING_HPP_
 
 #include <string>
+#include <memory_resource>
 #include "rosidl_runtime_c/u16string_functions.h"
 
 namespace rmw_cyclonedds_cpp
 {
 
 void u16string_to_wstring(
-  const rosidl_runtime_c__U16String & u16str, std::wstring & wstr);
+  const rosidl_runtime_c__U16String & u16str, std::pmr::wstring & wstr);
 
 bool wstring_to_u16string(
-  const std::wstring & wstr, rosidl_runtime_c__U16String & u16str);
+  const std::pmr::wstring & wstr, rosidl_runtime_c__U16String & u16str);
 
-void u16string_to_wstring(const std::u16string & u16str, std::wstring & wstr);
+void u16string_to_wstring(const std::pmr::u16string & u16str, std::pmr::wstring & wstr);
 
-bool wstring_to_u16string(const std::wstring & wstr, std::u16string & u16str);
+bool wstring_to_u16string(const std::pmr::wstring & wstr, std::pmr::u16string & u16str);
 
 }  // namespace rmw_cyclonedds_cpp
 
