diff --git a/rosidl_typesupport_fastrtps_c/include/rosidl_typesupport_fastrtps_c/wstring_conversion.hpp b/rosidl_typesupport_fastrtps_c/include/rosidl_typesupport_fastrtps_c/wstring_conversion.hpp
index ce85914..f8981af 100644
--- a/rosidl_typesupport_fastrtps_c/include/rosidl_typesupport_fastrtps_c/wstring_conversion.hpp
+++ b/rosidl_typesupport_fastrtps_c/include/rosidl_typesupport_fastrtps_c/wstring_conversion.hpp
@@ -20,27 +20,48 @@
 #include "rosidl_runtime_c/u16string.h"
 #include "rosidl_typesupport_fastrtps_c/visibility_control.h"
 
+#ifdef __linux__
+#include <experimental/memory_resource>
+#include <experimental/vector>
+#include <experimental/string>
+#else
+#include <memory_resource>
+#include <vector>
+#include <string>
+#endif
+
+#ifdef __linux__
+namespace std::pmr
+{
+    using string = std::experimental::pmr::string;
+    template <class _ValueT>
+    using vector = std::experimental::pmr::vector<_ValueT>;
+    using wstring = std::experimental::pmr::wstring;
+    using u16string = std::experimental::pmr::u16string;
+}
+#endif
+
 namespace rosidl_typesupport_fastrtps_c
 {
 
-/// Convert a `rosidl_runtime_c__U16String` into a std::wstring
+/// Convert a `rosidl_runtime_c__U16String` into a std::pmr::wstring
 /**
  * \param[in] u16str The 16-bit character string to copy from.
- * \param[in,out] wstr The std::wstring to copy to.
+ * \param[in,out] wstr The std::pmr::wstring to copy to.
  */
 ROSIDL_TYPESUPPORT_FASTRTPS_C_PUBLIC
 void u16string_to_wstring(
-  const rosidl_runtime_c__U16String & u16str, std::wstring & wstr);
+  const rosidl_runtime_c__U16String & u16str, std::pmr::wstring & wstr);
 
-/// Convert a std::wstring into a `rosidl_runtime_c__U16String`.
+/// Convert a std::pmr::wstring into a `rosidl_runtime_c__U16String`.
 /**
- * \param[in] wstr The std::wstring to copy from.
+ * \param[in] wstr The std::pmr::wstring to copy from.
  * \param[in,out] u16str The u16string to copy to.
  * \return true if resizing u16str and assignment succeeded, otherwise false.
  */
 ROSIDL_TYPESUPPORT_FASTRTPS_C_PUBLIC
 bool wstring_to_u16string(
-  const std::wstring & wstr, rosidl_runtime_c__U16String & u16str);
+  const std::pmr::wstring & wstr, rosidl_runtime_c__U16String & u16str);
 
 }  // namespace rosidl_typesupport_fastrtps_c
 
diff --git a/rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em b/rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em
index 7ca24f9..2f317c2 100644
--- a/rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em
+++ b/rosidl_typesupport_fastrtps_c/resource/msg__type_support_c.cpp.em
@@ -201,7 +201,7 @@ if isinstance(type_, AbstractNestedType):
       cdr << str->data;
     }
 @[    elif isinstance(member.type.value_type, AbstractWString)]@
-    std::wstring wstr;
+    std::pmr::wstring wstr;
     for (size_t i = 0; i < size; ++i) {
       const rosidl_runtime_c__U16String * str = &array_ptr[i];
       if (str->capacity == 0 || str->capacity <= str->size) {
@@ -246,7 +246,7 @@ if isinstance(type_, AbstractNestedType):
     }
     cdr << str->data;
 @[  elif isinstance(member.type, AbstractWString)]@
-    std::wstring wstr;
+    std::pmr::wstring wstr;
     rosidl_typesupport_fastrtps_c::u16string_to_wstring(ros_message->@(member.name), wstr);
     cdr << wstr;
 @[  elif isinstance(member.type, BasicType) and member.type.typename == 'boolean']@
@@ -342,7 +342,7 @@ else:
       }
     }
 @[    elif isinstance(member.type.value_type, AbstractWString)]@
-    std::wstring wstr;
+    std::pmr::wstring wstr;
     for (size_t i = 0; i < size; ++i) {
       auto & ros_i = array_ptr[i];
       if (!ros_i.data) {
@@ -396,7 +396,7 @@ else:
     if (!ros_message->@(member.name).data) {
       rosidl_runtime_c__U16String__init(&ros_message->@(member.name));
     }
-    std::wstring wstr;
+    std::pmr::wstring wstr;
     cdr >> wstr;
     bool succeeded = rosidl_typesupport_fastrtps_c::wstring_to_u16string(wstr, ros_message->@(member.name));
     if (!succeeded) {
diff --git a/rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/wstring_conversion.hpp b/rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/wstring_conversion.hpp
index 35bc7c3..d42347f 100644
--- a/rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/wstring_conversion.hpp
+++ b/rosidl_typesupport_fastrtps_cpp/include/rosidl_typesupport_fastrtps_cpp/wstring_conversion.hpp
@@ -16,8 +16,28 @@
 #define ROSIDL_TYPESUPPORT_FASTRTPS_CPP__WSTRING_CONVERSION_HPP_
 
 #include <rosidl_typesupport_fastrtps_cpp/visibility_control.h>
+#include <string>
 
+#ifdef __linux__
+#include <experimental/memory_resource>
+#include <experimental/vector>
+#include <experimental/string>
+#else
+#include <memory_resource>
+#include <vector>
 #include <string>
+#endif
+
+#ifdef __linux__
+namespace std::pmr
+{
+ using string = std::experimental::pmr::string;
+ template <class _ValueT>
+ using vector = std::experimental::pmr::vector<_ValueT>;
+ using wstring = std::experimental::pmr::wstring;
+ using u16string = std::experimental::pmr::u16string;
+}
+#endif
 
 namespace rosidl_typesupport_fastrtps_cpp
 {
@@ -28,7 +48,7 @@ namespace rosidl_typesupport_fastrtps_cpp
  * \param[in,out] wstr The std::wstring to copy to.
  */
 ROSIDL_TYPESUPPORT_FASTRTPS_CPP_PUBLIC
-void u16string_to_wstring(const std::u16string & u16str, std::wstring & wstr);
+void u16string_to_wstring(const std::pmr::u16string & u16str, std::pmr::wstring & wstr);
 
 /// Convert a std::wstring into a std::u16string.
 /**
@@ -37,7 +57,7 @@ void u16string_to_wstring(const std::u16string & u16str, std::wstring & wstr);
  * \return true if resizing u16str and assignment succeeded, otherwise false.
  */
 ROSIDL_TYPESUPPORT_FASTRTPS_CPP_PUBLIC
-bool wstring_to_u16string(const std::wstring & wstr, std::u16string & u16str);
+bool wstring_to_u16string(const std::pmr::wstring & wstr, std::pmr::u16string & u16str);
 
 }  // namespace rosidl_typesupport_fastrtps_cpp
 
diff --git a/rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em b/rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em
index 39dc452..fbef3e6 100644
--- a/rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em
+++ b/rosidl_typesupport_fastrtps_cpp/resource/msg__type_support.cpp.em
@@ -102,7 +102,7 @@ cdr_serialize(
     cdr << ros_message.@(member.name);
 @[      else]@
 @[        if isinstance(member.type.value_type, AbstractWString)]@
-    std::wstring wstr;
+    std::pmr::wstring wstr;
 @[        end if]@
     for (size_t i = 0; i < @(member.type.size); i++) {
 @[        if isinstance(member.type.value_type, NamespacedType)]@
@@ -134,7 +134,7 @@ cdr_serialize(
     }
 @[        else]@
 @[            if isinstance(member.type.value_type, AbstractWString)]@
-    std::wstring wstr;
+    std::pmr::wstring wstr;
 @[            end if]@
     for (size_t i = 0; i < size; i++) {
 @[            if isinstance(member.type.value_type, BasicType) and member.type.value_type.typename == 'boolean']@
@@ -162,7 +162,7 @@ cdr_serialize(
   cdr << static_cast<wchar_t>(ros_message.@(member.name));
 @[  elif isinstance(member.type, AbstractWString)]@
   {
-    std::wstring wstr;
+    std::pmr::wstring wstr;
     rosidl_typesupport_fastrtps_cpp::u16string_to_wstring(ros_message.@(member.name), wstr);
     cdr << wstr;
   }
@@ -192,7 +192,7 @@ cdr_deserialize(
     cdr >> ros_message.@(member.name);
 @[      else]@
 @[        if isinstance(member.type.value_type, AbstractWString)]@
-    std::wstring wstr;
+    std::pmr::wstring wstr;
 @[        end if]@
     for (size_t i = 0; i < @(member.type.size); i++) {
 @[        if isinstance(member.type.value_type, NamespacedType)]@
@@ -223,7 +223,7 @@ cdr_deserialize(
     }
 @[        else]@
 @[            if isinstance(member.type.value_type, AbstractWString)]@
-    std::wstring wstr;
+    std::pmr::wstring wstr;
 @[            end if]@
     for (size_t i = 0; i < size; i++) {
 @[            if isinstance(member.type.value_type, BasicType) and member.type.value_type.typename == 'boolean']@
@@ -266,7 +266,7 @@ cdr_deserialize(
   }
 @[  elif isinstance(member.type, AbstractWString)]@
   {
-    std::wstring wstr;
+    std::pmr::wstring wstr;
     cdr >> wstr;
     bool succeeded = rosidl_typesupport_fastrtps_cpp::wstring_to_u16string(wstr, ros_message.@(member.name));
     if (!succeeded) {
diff --git a/rosidl_typesupport_fastrtps_cpp/src/wstring_conversion.cpp b/rosidl_typesupport_fastrtps_cpp/src/wstring_conversion.cpp
index 387359d..c27adcc 100644
--- a/rosidl_typesupport_fastrtps_cpp/src/wstring_conversion.cpp
+++ b/rosidl_typesupport_fastrtps_cpp/src/wstring_conversion.cpp
@@ -17,7 +17,7 @@
 namespace rosidl_typesupport_fastrtps_cpp
 {
 
-void u16string_to_wstring(const std::u16string & u16str, std::wstring & wstr)
+void u16string_to_wstring(const std::pmr::u16string & u16str, std::pmr::wstring & wstr)
 {
   wstr.resize(u16str.size());
   for (size_t i = 0; i < u16str.size(); ++i) {
@@ -25,7 +25,7 @@ void u16string_to_wstring(const std::u16string & u16str, std::wstring & wstr)
   }
 }
 
-bool wstring_to_u16string(const std::wstring & wstr, std::u16string & u16str)
+bool wstring_to_u16string(const std::pmr::wstring & wstr, std::pmr::u16string & u16str)
 {
   try {
     u16str.resize(wstr.size());
