diff --git a/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake b/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
index a6d8263..0087d9a 100644
--- a/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
+++ b/rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake
@@ -183,6 +183,14 @@ target_include_directories(${_target_name_lib}
   ${PythonExtra_INCLUDE_DIRS}
 )
 
+find_library(UTIL_LIBRARY NAMES util)
+if(NOT UTIL_LIBRARY)
+    message(FATAL_ERROR "libutil not found")
+endif()
+target_link_libraries(${_target_name_lib}
+    ${UTIL_LIBRARY}
+)
+
 # Check if numpy is in the include path
 find_file(_numpy_h numpy/numpyconfig.h
   PATHS ${PythonExtra_INCLUDE_DIRS}
